### DISCUSSION TITLE
Streamline local dev of plv8 functions

### DISCUSSION THREAD

**Discussion by wiverson** [supabase/supabase#11146] [Feature Requests]
# Feature request

Add support for something like https://github.com/divyenduz/plv8ify to the local dev environment build process. 

## Is your feature request related to a problem? Please describe.

I would like to be able to write ordinary TypeScript files and have them get compiled and added to the project as plv8 functions in a similar manner. Ideally, I could then debug those functions with proper source maps back to the original function declaration.

## Describe the solution you'd like

Either locally or in the cloud, create TypeScript functions with [decorators](https://www.typescriptlang.org/docs/handbook/decorators.html) for key metadata along the lines of how pl/java uses an [annotation](https://tada.github.io/pljava/pljava-api/apidocs/org.postgresql.pljava/org/postgresql/pljava/annotation/Function.html) to provide pl needed information.

Ideally, then I could attach a [debugger](https://github.com/supabase/supabase/issues/11138) and debug my function as well.

## Describe alternatives you've considered

Right now implementing PostgREST functions requires a fair amount of awareness of a lot of Postgres features as well as a working familiarity with the $$ syntax. It works but it's very clunky. No IDE support means no formatting, no syntax highlighting, no debugging - it's basically just notepad development with a very clunky syntax.

## Additional context

I assume(?) that most Supabase developers are coming to the platform as JavaScript/TypeScript developers first, and then adding in SQL skills as needed to get the job done. These improvements would make JavaScript development for server-side functions first-class citizens.

**Comment by davepar** [comment]
I agree it would be nice to simplify/improve plv8 development. I'm imaging something much simpler:

* One TS function per file, no imports.
* The function is converted to plv8 by compiling TS->JS and adding `create function fn_name(<params>) returns ret_type as $$<code>$$ language plv8;`
* The TS param and return types are converted to appropriate Postgres types.
* Somehow allow specifying the schema (if not public) and adding optionally adding "security definer".
* Maybe a supabase CLI command that compiles and executes the `drop function` & `create function` in Postgres.

This would enormously simplify the process of writing and updating plv8 functions, and allow for writing tests against the TS code. And as a bonus, mocking data for `plv8.execute()` and other fns for testing.

**Reply by davepar** [reply]
I'm starting to work on the bullet points I mentioned above. I'm tired of running into errors in my plv8 code that would have been easily caught by the TS compiler. I should have it on github in the next day or so.

**Reply by wiverson** [reply]
LMK if/when you post something and I'd be happy to give it a spin.  :)

**Reply by davepar** [reply]
I have something working.  It does everything I mentioned above except automatically integrating with Supabase. That step is just: `psql -h <dbhost> -p 5432 -d postgres -U postgres -f <filename>`.

I also added lint and prettier (from Google TS Style, which is a bit non-standard but easy to install). The type checking works great and found some potential issues in the functions I previously wrote.

Let me know if it works for you: https://github.com/Davepar/plv8-converter

**Comment by davepar** [comment]
I switched things up and made an  NPM package for compiling Typescript to PLV8 functions: https://www.npmjs.com/package/plv8ts
